### DISCUSSION TITLE
Location of definition of draw functions?

### DISCUSSION THREAD

**Discussion by goyalyashpal** [networkx/networkx#6638] [Q&A]
### Summaries/Titles
- `ms-python.vscode-pylance` shows errors with networkx.draw related functions 
(Python LS/intellisense for VSCode) 
- draw not known member of networkx module - Pylance
- Location of definition of draw functions not known

### Current Behavior

3. "draw_networkx" is not a known member of module "networkx" Pylance [reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues) [Ln 9, Col 10]
(function) draw_networkx: Unknown
4. No definition found for `draw_networkx`

### Expected Behavior

Documentation for the function is shown

### Steps to Reproduce

1. Open VSCodium with `ms-python.python` extension enabled
2. Save this following code to a file with python valid file name & extension and open it with vscodium > ensure language mode is set to Python
3. This requires turning on the "Python Type Checking Mode" from status bar (default is off). This will show as Set to "basic".
    1. Open the Problems Panel `ctrl + shift + m` or
    2. Hover over or click on `draw` (line 9) and then do `ctrl + shift + p` Command Palette > type `hover` > select "Show definition preview hover"
4. Click on `draw` (line 9) > press `f12` "Go to definition"
 

```python
import networkx
import matplotlib.pyplot as plt

edges = [(1,2),(2,3),(2,4)]

nxgraph = networkx.DiGraph()
nxgraph.add_edges_from(edges)

# networkx.draw_networkx(nxgraph)
networkx.draw(nxgraph)
plt.show()

```

### Environment

Python: 3.11.2 (via `python --version`)
NetworkX: 3.1 (via `pip list | grep networkx`)
OS: Windows 10 22H2 x64

VSCodium: 1.77.0
Pylance extension `ms-python.vscode-pylance` : v2023.3.41
Python extension `ms-python.python` : v2023.6.0

**Comment by rossbar** [comment]
The problem arises from your IDE setup, not networkx. In this case the types are being inferred by pylance. I would first check that networkx is installed as you expect in your working environment. Once you've verified that, the next step would be to check the pylance issue tracker for similar issues - if not, then consider opening an issue there.

**Reply by goyalyashpal** [reply]
looking at the directory structure of other correctly inferred packages like `matplotlib` vs of `networkx`, i think Pylance is unable to look into the folders...

```console
$ ls -cF --group-directories-first "...\Python\Python311\Lib\site-packages\networkx"
__pycache__/  tests/      linalg/      drawing/  algorithms/  relabel.py       exception.py       conftest.py
utils/        readwrite/  generators/  classes/  __init__.py  lazy_imports.py  convert_matrix.py  convert.py

$ ls -cF --group-directories-first "...\Python\Python311\Lib\site-packages\matplotlib"
__pycache__/  _ttconv.cp311-win_amd64.pyd*  _color_data.py          texmanager.py   markers.py   container.py
_api/         _tight_bbox.py                _constrained_layout.py  table.py        mathtext.py  contour.py
tri/          _tight_layout.py              _cm.py                  streamplot.py   lines.py     colorbar.py
tests/        _qhull.cp311-win_amd64.pyd*   _cm_listed.py           spines.py       legend.py    colors.py
...
```

**Reply by goyalyashpal** [reply]
is networkx being correctly inferred in your environment? if yes, then what LS are u using with which editor? (this would be required even if for filing issues)

**Comment by goyalyashpal** [comment]
thanks for the pointers @ rossbar , [looking at the Pylance issues](https://github.com/microsoft/pylance-release/issues?q=is%3Aissue+networkx+completion), i found this:

> ```python
> networkx.readwrite.gpickle.<tab>
> ```
> The reason the completions are not showing is because networkx does not explicitly export the `readwrite` module or the `gpickle` module. if you look at the top level `__init.py__` you will notice that `readwrite` is not exported like they have done with `from networkx.testing.test import run as test` 
>
> Therefore, you will need to explicitly import the gpickle module like `import networkx.readwrite.gpickle as gp` and then `gp.` should give you completions.
>
> @judej at https://github.com/microsoft/pylance-release/issues/682#issuecomment-738334645

_(Sorry judej for mentioning, my reasoning is that ur participation here with networkx team may be more direct)_

**Comment by rossbar** [comment]
I agree with judej and don't think there's anything to be done here. Also note that the packages in question have since been removed from networkx as of v3.0. I will close this as resolved!

**Reply by goyalyashpal** [reply]
@_@

**Reply by goyalyashpal** [reply]
> Also note that the packages in question have since been removed from networkx as of v3.0. I will close this as resolved!

what are the "packages in question" which are "removed from nx as of v3.0" --> draw is removed from nx??

https://networkx.org/documentation/stable/reference/drawing.html#drawing

it still has no mention of removal or anything?

well, okay, i don't care either. thanks for providing info.
